first pass at key uniqueness checking

nickeubank opened the pull request:
Half of #1412 

Can anyone advise on how to handle the `on` argument for stranger cases (like dictionaries)? Looks like there's a system in place (`OnType`), but can't quite figure out.

nickeubank commented:
OK, think I fixed the `OnType` issue...

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -247,6 +273,11 @@ Join two `DataFrame` objects
   if `true`, duplicate names will be suffixed with `_i`
   (`i` starting at 1 for the first duplicate).
 
+* `check_uniqueness` : checks for whether merge keys are unique before merging.
```
I suggest something like (feel free to improve it if you have ideas):
"whether to check that columns passed as the `on` argument define unique keys in each input data frame. Can be a tuple or a pair, with the first element indicating whether to run check for `df1` and the second element for `df2`. By default no check is performed."

Also the name of the argument isn't great, as we already have `makeunique` which refers to column names. Not sure what's the best name. Maybe `validate` as in Pandas? I agree it's not very specific.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -280,7 +311,9 @@ join(name, job2, on = :ID => :identifier)
                    on::Union{<:OnType, AbstractVector{<:OnType}} = Symbol[],
-                   kind::Symbol = :inner, makeunique::Bool=false)
+                   kind::Symbol = :inner, makeunique::Bool=false,
+                   check_uniqueness::Tuple{Bool,Bool}=(false, false))
```
Would be nice to also allow `Pair{Bool,Bool}` as we do for `on`.

Also don't add space below.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -212,8 +212,34 @@ function update_row_maps!(left_table::AbstractDataFrame,
     return to_bimap(left_ixs), to_bimap(leftonly_ixs), to_bimap(right_ixs), to_bimap(rightonly_ixs)
 end
 
+function check_uniqueness_func(check_uniqueness, joiner)
```
Given that it's only called from one place, there's no point in defining  a separate function.

nickeubank replied:
I know I don't have to, but there is a point: good decomposition / style! It's a lot of clutter to leave in the core function definition, no?

nalimilan replied:
I think with my suggestion you should be able to make it shorter.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -212,8 +212,34 @@ function update_row_maps!(left_table::AbstractDataFrame,
 
+function check_uniqueness_func(check_uniqueness, joiner)
+
+    if check_uniqueness == (true, false) && any(nonunique(joiner.dfl, joiner.left_on))
```
I'd simplify this code by handling `df1` and `df2` separately (if `check_uniqueness` is false then just do as if the check was OK), and only then have an `if` with three branches determining the error message to print.

Since you compute `nonunique`, it would also be nice to give the name of the first row for which the problem was found. It's very useful when trying to find out the problem.

nickeubank replied:
I split it out to minimize computational costs. This way we don't ever run a `nonunique` unnecessarily. If I do `df1` and `df2` up front, we run it on both even if we only need to check one. 

> Since you compute nonunique, it would also be nice to give the name of the first row for which the problem was found. It's very useful when trying to find out the problem.

Good idea. Will put in error message.

nalimilan replied:
Yes but you can do something like:
```julia
left_valid = check_uniqueness[1] ? any(nonunique(joiner.dfl, joiner.left_on)) : true
right_valid = check_uniqueness[2] ? any(nonunique(joiner.dfr, joiner.right_on)) : true
```

nickeubank replied:
ok

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,23 @@ module TestJoin
     end
+
+    @testset "test checks of merge key uniqueness" begin
+
```
No need for all these empty lines here nor below (see rest of the file for examples).

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,23 @@ module TestJoin
+    @testset "test checks of merge key uniqueness" begin
+
+        @test_throws ArgumentError join(name, job, on=:ID,
+                     check_uniqueness=(false, true))
```
Align arguments with parentheses (or maybe they even fit on the previous line?).

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,23 @@ module TestJoin
+                      check_uniqueness=(true, true))
+
+        @test join(name, job, on=:ID, check_uniqueness=(true, false)) ==  inner
+        @test join(name, job, on = :ID, kind = :inner, check_uniqueness=(false, false)) == inner
```
No spaces around `=` for consistency.

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,23 @@ module TestJoin
             [Union{Int, Missing}, CS, Union{Int, Missing}]
     end
+
+    @testset "test checks of merge key uniqueness" begin
```
Can you also test what happens when the keys contain missing values? What are we supposed to do in that case (and what does Stata do)?

nickeubank replied:
test added

nickeubank commented:
fixed, rebased, and squashed

bkamins commented:
I am not sure if my comments are visible - at least I do not see them (sorry - no experience in reviewing yet). I will sum up what I suggested other than @nalimilan:

1. clearly write that `isequal` is used to check for uniqueness in the documentation.
2. Add tests not only for `missing` but also for `NaN` and `0.0`/`-0.0`.

Sorry for the mess

nalimilan commented:
> Is that what's used for nonunique? I'm just using nonunique. Is it better to cite that?

Yes, it uses the same grouping code as `groupby`, which is based on the `Dict` code, which in turn uses `isequal`. It's probably clearer to mention `isequal`, which people are more likely to be familiar with.

bkamins commented:
This is the point; `isequal` is a standard contract defined in Base.

nickeubank commented:
ok, think that hits all the issues raised!

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -247,6 +248,12 @@ Join two `DataFrame` objects
   if `true`, duplicate names will be suffixed with `_i`
   (`i` starting at 1 for the first duplicate).
 
+* `validate` : whether to check that columns passed as the on argument
```
`` `on` ``

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -247,6 +248,12 @@ Join two `DataFrame` objects
+   define unique keys in each input data frame. Can be a tuple or a pair,
+   with the first element indicating whether to run check for `df1` and
+   the second element for `df2`. By default no check is performed.
+   Merge key equality it tested using `isequal`.
```
"Merge" isn't the terminology we use. I'd just add this mention this after "in each input data frame", like this: ``(according to [`isequal`](@ref)) ``.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -290,6 +298,26 @@ function Base.join(df1::AbstractDataFrame,
+        first_error = findfirst(nonunique(joiner.dfr, joiner.right_on))
+        throw(ArgumentError("Merge key(s) in df2 are not unique. " *
+                            "First duplicate at row $first_error"))
+    elseif left_invalid && right_invalid
```
If you put this branch first, you can avoid the `!left_invalid` and `!right_invalid` checks.

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,54 @@ module TestJoin
+        @test join(name, job, on=:ID, validate=(true, false)) ==  inner
+        @test join(name, job, on=:ID, kind=:inner, validate=(false, false)) == inner
+
+        # Make sure ok with various weird characters
```
These are not "characters". And "weird" could just be "special".

Below, it would be useful to include a `0` in the values of `ID`, to check that it's treated as different from `-0.0`. Then you can drop `0.0` from the list here. Without this, the code doesn't actually test that `0.0` and `-0.0` are considered as different.

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,54 @@ module TestJoin
+        # Make sure ok with various weird characters
+        for weird in [missing, NaN, 0.0, -0.0]
+            name_w_missing = DataFrame(ID = [1, 2, 3, weird],
+                                       Name = Union{String, Missing}["John Doe", "Jane Doe",
```
`Union{String, Missing}` isn't useful here, right? Same below.

nalimilan replied:
These type specifications have been preserved in existing tests when moving from `DataArray`, to make sure we still test such arrays. But in new code it's generally not needed.

nickeubank replied:
k

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,54 @@ module TestJoin
+                                                                     "Joe Blogs", "Maria Tester"])
+            @test join(name_w_missing, job, on=:ID, validate=(true, false)) ==  inner
+
+            # Make sure duplicated missings still an exception
```
Comment and variable names should be adjusted since you test other values than missing.

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,54 @@ module TestJoin
+
+    @testset "test checks of merge key uniqueness" begin
+        @test_throws ArgumentError join(name, job, on=:ID,
+                                        validate=(false, true))
```
Better put this on the previous line, it's well below 92 chars.

nalimilan replied:
Julia's CONTRIBUTING.md mentions 92 chars, that's usually the target we try to follow.

nickeubank replied:
wow that is so much more breathing room

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,54 @@ module TestJoin
+                                 Name = ["John Doe", "Jane Doe", "Joe Blogs", missing, missing, missing],
+                                 Job = ["Lawyer", missing, missing, "Doctor", "Florist",  "Farmer"])
+
+         @test isequal(join(name_multi, job_multi, on=[:ID1, :ID2], kind=:outer, validate=(true, false)),
```
Use `≅`.

nickeubank commented:
ok I think that covers everything. Thanks for all the help @nalimilan @bkamins

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -290,6 +298,26 @@ function Base.join(df1::AbstractDataFrame,
+    if left_invalid && right_invalid
+        first_error_df1 = findfirst(nonunique(joiner.dfl, joiner.left_on))
+        first_error_df2 = findfirst(nonunique(joiner.dfr, joiner.right_on))
+        throw(ArgumentError("Merge key(s) are not unique in df1 or df2. " *
```
"or" -> "and"?

nalimilan commented on test/join.jl:
```diff
@@ -578,4 +578,64 @@ module TestJoin
+
+        @test join(name_w_zeros, name_w_zeros, on=:ID, validate=(true, true)) ≅ name_w_zeros2
+
+         # Check for multiple-column merge keys
```
Incorrect alignment.

nickeubank commented:
any more open issues here?

bkamins commented on src/abstractdataframe/join.jl:
```diff
@@ -290,6 +298,26 @@ function Base.join(df1::AbstractDataFrame,
     joiner = DataFrameJoiner(df1, df2, on)
 
+    # Check merge key validity
+    left_invalid = validate[1] ? any(nonunique(joiner.dfl, joiner.left_on)) : false
```
I do not remember if it was discussed but `nonunique(joiner.dfl, joiner.left_on)` and `nonunique(joiner.dfr, joiner.right_on)` could be saved in a variable to avoid making the call again on error.

bkamins commented on src/abstractdataframe/join.jl:
```diff
@@ -290,6 +298,26 @@ function Base.join(df1::AbstractDataFrame,
+    if left_invalid && right_invalid
+        first_error_df1 = findfirst(nonunique(joiner.dfl, joiner.left_on))
+        first_error_df2 = findfirst(nonunique(joiner.dfr, joiner.right_on))
+        throw(ArgumentError("Merge key(s) are not unique in either df1 or df2. " *
```
Should it rather be `not unique in df1 and df2` as we know that there are duplicates in both `DataFrame`s?

bkamins commented:
I have left some small comments (but in general PR looks good to merge).

nickeubank commented:
Rebased and corrected message. Open to suggestion on `nonunique` issue noted above. 

Note this PR now includes a patch for #1424  -- I realized on change we made resulted in modification of the input datasets. In particular, the correction is now at Lines 307-308 (using `hcat` instead of direct column insertion). Tests to ensure datasets not modified added as well.

bkamins commented:
I am OK to leave `nonunique` as is, so it is good to merge for me. Thanks!

bkamins commented:
In the future the 307-308 patch would be better to be a separate commit (or separate PR) to keep history or changes.

nickeubank commented:
@bkamins Oh, ok -- sorry about that!  It came up here (that error broke the tests for this PR) so I just patched, but in the future I'll keep them separate.

nalimilan commented:
Yes, please move that commit to a different PR, we already have enough changes on our plate here.

nalimilan commented on test/join.jl:
```diff
@@ -578,14 +578,17 @@ module TestJoin
         @test eltypes(join(l, r, on=:b, kind=:outer, makeunique=true)) ==
             [Union{Int, Missing}, CS, Union{Int, Missing}]
     end
-
```
Keep this line, and add a break before the `@testset` you added.

nalimilan commented on test/join.jl:
```diff
@@ -578,14 +578,17 @@ module TestJoin
                                     Name = ["John Doe", "Jane Doe", "Jane Doe", "Joe Blogs", missing],
                                     Job = ["Lawyer", "Doctor", "Florist", missing, "Farmer"],
                                     _merge = ["both", "both", "both", "left_only", "right_only"])
+        pre_join_name = copy(name)
```
I don't understand why this is needed.

nickeubank commented:
OK, split out in #1434 . Note these tests will fail until that patch is merged.

nalimilan commented:
Thanks!
